pkg/webhook/clustersetbinding/validating_webhook.go

```diff
@@ -0,0 +1,145 @@
+		admissionSpec.Resource.Resource != "managedclustersetbindings" {
+		status.Allowed = true
+		return status
+	}
+
+	switch admissionSpec.Operation {
+	case admissionv1beta1.Create, admissionv1beta1.Update:
+		return a.validateRequest(admissionSpec)
+	default:
+		status.Allowed = true
+		return status
+	}
+}
+
+// Initialize is called by generic-admission-server on startup to setup initialization that ManagedClusterSetBinding webhook needs.
+func (a *ManagedClusterSetBindingValidatingAdmissionHook) Initialize(kubeClientConfig *rest.Config, stopCh <-chan struct{}) error {
+	var err error
+	a.kubeClient, err = kubernetes.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return err
+	}
+
+	a.clusterClient, err = clusterclient.NewForConfig(kubeClientConfig)
+	return err
+}
+
+// validateCreateRequest validates the create/update operation of ManagedClusterSetBinding
+func (a *ManagedClusterSetBindingValidatingAdmissionHook) validateRequest(request *admissionv1beta1.AdmissionRequest) *admissionv1beta1.AdmissionResponse {
+	status := &admissionv1beta1.AdmissionResponse{}
+
+	binding := &clusterv1alpha1.ManagedClusterSetBinding{}
+	if err := json.Unmarshal(request.Object.Raw, binding); err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	// check if target cluster set is not specified
+	if len(binding.Spec.ClusterSet) == 0 {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: "Spec.ClusterSet is required",
+		}
+		return status
+	}
+
+	// check if the request user has permission to bind the target cluster set
+	return a.allowBindingToClusterSet(binding.Spec.ClusterSet, request.UserInfo)
+}
+
+// allowBindingToClusterSet checks if the user is able to bind a cluster set
+func (a *ManagedClusterSetBindingValidatingAdmissionHook) allowBindingToClusterSet(clusterSetName string, userInfo authenticationv1.UserInfo) *admissionv1beta1.AdmissionResponse {
+	status := &admissionv1beta1.AdmissionResponse{}
+
+	extra := make(map[string]authorizationv1.ExtraValue)
+	for k, v := range userInfo.Extra {
+		extra[k] = authorizationv1.ExtraValue(v)
+	}
+
+	sar := &authorizationv1.SubjectAccessReview{
+		Spec: authorizationv1.SubjectAccessReviewSpec{
+			User:   userInfo.Username,
+			UID:    userInfo.UID,
+			Groups: userInfo.Groups,
+			Extra:  extra,
+			ResourceAttributes: &authorizationv1.ResourceAttributes{
+				Group:       "cluster.open-cluster-management.io",
+				Resource:    "managedclustersets",
+				Subresource: "bind",
+				Verb:        "get",
+				Name:        clusterSetName,
+			},
+		},
+	}
+	sar, err := a.kubeClient.AuthorizationV1().SubjectAccessReviews().Create(context.TODO(), sar, metav1.CreateOptions{})
```

Comment:
seems like this check should only be done if the value of clusterset is changing. if you're just adding an annotation or something, is there a reason to restrict access like this?

Reply:
Removed the permission checking when updating cluster set binding.